### PR TITLE
client: get site from go:embed

### LINKED ISSUE #1664: site: look into go:embed for the site resources

### PR DESCRIPTION
**Pull request by vctt94**
This PR closes #1664

For testing can run `go install` in dexc directory and make sure that the folder `site` isn't in the go executable's directory or the working directory.

### CODE REVIEW

**Comment by buck54321** [comment]
This looks pretty slick. How do we get the tests to pass?

**Comment by vctt94** [comment]
tests passing now.

With this PR we will need to run: `npm run build` before pushing to repo to make sure we have the last site build.

**Comment by chappjc** [comment]
Just to reiterate the discussion in matrix, we need a way to allow CI to function without checking in the build output (dist). Perhaps just some placeholder file in the dist folder like `touch dist/goembed` so go build doesn't error.  Possibly it could be solved by tweaking the embed line.  It's just 1.3 MB whenever js/css modified, and that the build outputs really don't belong in source control, just the source.

The lack of the expected files in either (a) embed.FS OR (b) the actual file system, should result in a runtime error or just a 404 on the webserver as it would now, not a compile-time error.

**Comment by chappjc** [comment]
The other obvious solution is to merge the CI tasks so that the JS build and Go build are in the same runner.  That would be a bummer though because the build matrix is unnecessarily large in that case.

**Review by buck54321 — COMMENTED** [review]
A few changes, please.

1) Don't check in the **dist** files.
1) Let's put some empty dummy files (*entry.js* and *style.css*) in **client/webserver/** for the **/dist** files. That'll get us past the tests.

```Go
//go:embed entry.js
jsFile embed.FS

//go:embed style.css
cssFile embed.FS
```

3) We can recognize the empty files in `New`, and use the old search.
3) In the future, the packaging script can hot swap the compiled filed during the build. No need to do that in the PR.

**Comment by chappjc** [comment]
> Or maybe @chappjc's is right about just using an empty directory

What I meant was this:

```
$  pwd
<repo>/client/webserver
$  mkdir -p site/dist
$  touch site/dist/goembed
```

The Go binary will build now.  For releases, we can check in the final `dist` folder contents to people can do stuff like
`go install decred.org/dcrdex@release-v0.5.0` and it will use the dist folder that is on that branch.

**Comment by chappjc** [comment]
This is very close.   I'm going to suggest a diff in a few.

**Comment by chappjc** [comment]
Here's what I think we should change: https://github.com/decred/dcrdex/commit/0c6cfc1da8f47766dd584259799d4e731902c947
That diff applies on top of this PR currently.

When we start distributing Go binaries with these files embedded, when users upgrade they are likely to overwrite the old dexc(.exe) and leave the old site files.  As such, by default dexc needs to *only* use the embedded files, not first try disk nor fallback to disk.

The `--reload-html` arg is the same, but it is now the negation of "use embedded", meaning it will not use the embedded files and it will instead use serve disk _and_ always reload the html templates for each request.

After building, we can verify the embedded files with:

```
$  go list -f '{{ .EmbedFiles }}' decred.org/dcrdex/client/webserver
[site/dist/entry.js site/dist/entry.js.map site/dist/style.css site/dist/style.css.map site/src/font/SourceSansPro-Light.svg site/src/font/SourceSansPro-Light.ttf site/src/font/SourceSansPro-Light.woff site/src/font/SourceSansPro-Light.woff2 site/src/font/icomoon.svg site/src/font/icomoon.ttf site/src/font/icomoon.woff site/src/font/inconsolata-v15-latin-regular.svg site/src/font/inconsolata-v15-latin-regular.ttf site/src/font/inconsolata-v15-latin-regular.woff site/src/font/inconsolata-v15-latin-regular.woff2 site/src/font/source-sans-pro-semibold.svg site/src/font/source-sans-pro-semibold.ttf site/src/font/source-sans-pro-semibold.woff site/src/font/source-sans-pro-semibold.woff2 site/src/font/source-sans-pro-v9-latin-regular.svg site/src/font/source-sans-pro-v9-latin-regular.ttf site/src/font/source-sans-pro-v9-latin-regular.woff site/src/font/source-sans-pro-v9-latin-regular.woff2 site/src/img/candlestick_bttn.png site/src/img/coins/a.png site/src/img/coins/b.png site/src/img/coins/bch.png site/src/img/coins/btc.png site/src/img/coins/c.png site/src/img/coins/d.png site/src/img/coins/dcr.png site/src/img/coins/dextt.eth.png site/src/img/coins/doge.png site/src/img/coins/e.png site/src/img/coins/eth.png site/src/img/coins/f.png site/src/img/coins/g.png site/src/img/coins/h.png site/src/img/coins/i.png site/src/img/coins/j.png site/src/img/coins/k.png site/src/img/coins/l.png site/src/img/coins/ltc.png site/src/img/coins/m.png site/src/img/coins/mona.png site/src/img/coins/n.png site/src/img/coins/o.png site/src/img/coins/p.png site/src/img/coins/q.png site/src/img/coins/r.png site/src/img/coins/s.png site/src/img/coins/t.png site/src/img/coins/u.png site/src/img/coins/v.png site/src/img/coins/vtc.png site/src/img/coins/w.png site/src/img/coins/x.png site/src/img/coins/y.png site/src/img/coins/z.png site/src/img/coins/zec.png site/src/img/depth_bttn.png site/src/img/favicon.png site/src/img/logo_wide_dark_v1.svg site/src/img/logo_wide_v1.svg site/src/localized_html/en-US/bodybuilder.tmpl site/src/localized_html/en-US/dexsettings.tmpl site/src/localized_html/en-US/forms.tmpl site/src/localized_html/en-US/login.tmpl site/src/localized_html/en-US/markets.tmpl site/src/localized_html/en-US/order.tmpl site/src/localized_html/en-US/orders.tmpl site/src/localized_html/en-US/register.tmpl site/src/localized_html/en-US/settings.tmpl site/src/localized_html/en-US/wallets.tmpl site/src/localized_html/pl-PL/bodybuilder.tmpl site/src/localized_html/pl-PL/dexsettings.tmpl site/src/localized_html/pl-PL/forms.tmpl site/src/localized_html/pl-PL/login.tmpl site/src/localized_html/pl-PL/markets.tmpl site/src/localized_html/pl-PL/order.tmpl site/src/localized_html/pl-PL/orders.tmpl site/src/localized_html/pl-PL/register.tmpl site/src/localized_html/pl-PL/settings.tmpl site/src/localized_html/pl-PL/wallets.tmpl site/src/localized_html/pt-BR/bodybuilder.tmpl site/src/localized_html/pt-BR/dexsettings.tmpl site/src/localized_html/pt-BR/forms.tmpl site/src/localized_html/pt-BR/login.tmpl site/src/localized_html/pt-BR/markets.tmpl site/src/localized_html/pt-BR/order.tmpl site/src/localized_html/pt-BR/orders.tmpl site/src/localized_html/pt-BR/register.tmpl site/src/localized_html/pt-BR/settings.tmpl site/src/localized_html/pt-BR/wallets.tmpl site/src/localized_html/zh-CN/bodybuilder.tmpl site/src/localized_html/zh-CN/dexsettings.tmpl site/src/localized_html/zh-CN/forms.tmpl site/src/localized_html/zh-CN/login.tmpl site/src/localized_html/zh-CN/markets.tmpl site/src/localized_html/zh-CN/order.tmpl site/src/localized_html/zh-CN/orders.tmpl site/src/localized_html/zh-CN/register.tmpl site/src/localized_html/zh-CN/settings.tmpl site/src/localized_html/zh-CN/wallets.tmpl]
```

Default:

```
[DBG] WEB: Using embedded site resources.
[INF] WEB: Using localized HTML templates in <embedded>/site/src/localized_html/en-US
```

With `--reload-html`:

```
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/cmd/dexc/site
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/cmd/dexc/site
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/webserver/site
[INF] WEB: Located "site" folder at /home/jon/github/decred/dcrdex/client/webserver/site
[INF] WEB: Using localized HTML templates in /home/jon/github/decred/dcrdex/client/webserver/site/src/localized_html/en-US
```

With `--lang pt`:

```
[DBG] WEB: Using embedded site resources.
[INF] WEB: Using localized HTML templates in <embedded>/site/src/localized_html/pt-BR
```

With `--lang pt --reload-html`:

```
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/cmd/dexc/site
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/cmd/dexc/site
[DBG] WEB: Looking for site in /home/jon/github/decred/dcrdex/client/webserver/site
[INF] WEB: Located "site" folder at /home/jon/github/decred/dcrdex/client/webserver/site
[INF] WEB: Using localized HTML templates in /home/jon/github/decred/dcrdex/client/webserver/site/src/localized_html/pt-BR
```

Subsequent work should look at removing the `localized_html` folder from source control and instead generate the localized html templates at runtime.

File: pkg.sh [hunk 1]
```diff
@@ -33,25 +41,12 @@ GOOS=darwin GOARCH=amd64 go build -trimpath -o ../../../bin/dexc-darwin-amd64-v$
...
-popd
-
-rm -rf bin/site
-mkdir -p bin/site/src
-pushd client/webserver/site
-cp -R dist ../../../bin/site
-cp -R src/font src/html src/localized_html src/img ../../../bin/site/src
-popd
+echo "Files embedded in the Go webserver package:"
+go list -f '{{ .EmbedFiles }}' decred.org/dcrdex/client/webserver
```
Comment (chappjc) [inline pkg.sh:45]: Here's how we'll verify that the binaries have the intended site files embedded.

**Review by JoeGruffins — APPROVED** [review]
Working well.

File: pkg.sh [hunk 2]
```diff
@@ -13,6 +13,14 @@ mkdir -p bin/dexc-darwin-arm64-v${VER}
 
 export CGO_ENABLED=0
 
+# Generate the localized_html and build the webpack bundle prior to building the
+# webserver package, which embeds the files.
+pushd client/webserver/site
+go generate # should be a no-op
```
Comment (martonp) [inline pkg.sh:19]: Just wondering, what's the reason for this?

File: client/webserver/template.go [hunk 3]
```diff
@@ -26,27 +26,35 @@ type templates struct {
...
+		t.addErr = fmt.Errorf("no translation dictionary found for lang %q", lang)
+		return t
+	}
+	if embedded {
+		return t
 	}
+
+	// Make sure we have the expected directory structure.
+	if !folderExists(t.srcDir()) {
+		t.addErr = fmt.Errorf("reload-html set but source directory not found at %s", t.srcDir())
```
Comment (martonp) [inline client/webserver/template.go:55]: reload-html -> no-embed-site

**Review by martonp — COMMENTED** [review]
LGTM

**Review by buck54321 — CHANGES_REQUESTED** [review]
This is what I see after 1) switch to this branch, 2) `go build -tags lgpl`, 3) `./dex --simnet --rpc --log=debug`

image

If I 

```Go
go list -f '{{ .EmbedFiles }}' decred.org/dcrdex/client/webserver
```

I can see all of the image files.

On the front side, I'm seeing status 200 on the images requests, but the files appear to be empty.

File: pkg.sh (hunk 1 above)
Comment (chappjc) [reply pkg.sh:45]: I think I'm mistaken about the go list command.  e.g. I make the build with the dist dir and it embeds, then I can do the go list command and it shows the files.  Then I `rm -r dist` and do the list command again:

```
$  go list -f '{{ .EmbedFiles }}' decred.org/dcrdex/client/webserver
../webserver.go:81:13: pattern site/dist: no matching files found
```

So clearly that go list is not just checking the compiled webserver package in the go package cache or whatever, but something based on the current files.

In this pkg script it probably is doing what we expect, but in general it's not.

File: client/webserver/webserver.go [hunk 4]
```diff
@@ -366,35 +387,52 @@ func New(cfg *Config) (*WebServer, error) {
...
 
-func (s *WebServer) buildTemplates(lang string) error {
-	langs := make([]language.Tag, 0, 1)
-	dirs := make([]string, 0, 1)
-	var match string
-
-	htmlDir := filepath.Join(s.siteDir, "src", "localized_html")
-
-	fileInfos, err := os.ReadDir(htmlDir)
+// buildTemplates sets the templates in the WebServer's html field, which is are
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/webserver/webserver.go:398]: Fair enough.  In this case it's clear to say "prepares the HTML templates".
Although my stance on docs changes drastically between exported and unexported methods, where often the latter are helpers tor the programmer.

[the remark this answers is not in the document]
Comment (vctt94) [reply client/webserver/webserver.go:398]: I wasn't able to reproduce buck's issue with icons, but I did get a different error when registering:

image

did anyone else noticed it?

[the remark this answers is not in the document]
Comment (vctt94) [reply client/webserver/webserver.go:398]: Which browser buck? I am not being able to reproduce this issue. Have you just built with tags lgpl and run it? Here if I try:


image

File: pkg.sh (hunk 2 above)
Comment (chappjc) [reply pkg.sh:19]: Make sure the templates in localized_html correspond to the sources.  CI makes us check them in presently so it's likely a no-op in this script if it's being run on an unmodified git revision.

**Comment by chappjc** [comment]
I managed to repro the "empty" files.  For me it randomly made the big logo (logo_wide_dark_v1.svg) blank.  

image

Point the browser directly at it:

image

The reason seems to be a MIME issue:

image

(png instead of svg).

Not clear why it only sometimes happens, but it's gotta be related to content type so we know what do debug I think.

**Review by vctt94 — COMMENTED** [review]
Working as properly now. 

Thanks for the help with this work @chappjc.

**Comment by chappjc** [comment]
Np.  Sorry for rushing it as a late addition to 0.5

**Review by buck54321 — APPROVED** [review]

**Pull request merged**